Orange/widgets/utils/scaling.py

```diff
@@ -30,53 +31,59 @@ def rescale_data(self):
 
     def _compute_domain_data_stat(self):
         stt = self.domain_data_stat = \
-            getCached(self.data, DomainBasicStats, (self.data,))
-        for index in range(len(self.domain)):
-            attr = self.domain[index]
+            getCached(self.data, DomainBasicStats, (self.data, True))
+        domain = self.domain
+        for attr in domain.attributes + domain.class_vars + domain.metas:
             if attr.is_discrete:
                 self.attr_values[attr] = [0, len(attr.values)]
             elif attr.is_continuous:
-                self.attr_values[attr] = [stt[index].min, stt[index].max]
+                self.attr_values[attr] = [stt[attr].min, stt[attr].max]
 
     def _compute_scaled_data(self):
         data = self.data
         # We cache scaled_data and validArray to share them between widgets
         cached = getCached(data, "visualizationData")
         if cached:
-            self.original_data, self.scaled_data, self.valid_data_array = cached
+            self.data, self.scaled_data, self.valid_data_array = cached
             return
 
         Y = data.Y if data.Y.ndim == 2 else np.atleast_2d(data.Y).T
-        self.original_data = np.hstack((data.X, Y)).T
-        self.scaled_data = no_jit = self.original_data.copy()
-        self.valid_data_array = ~np.isnan(no_jit)
-        for index in range(len(data.domain)):
-            attr = data.domain[index]
+        if np.any(data.metas):
+            all_data = (data.X, Y, data.metas)
+        else:
+            all_data = (data.X, Y)
+        all_data = np.hstack(all_data).T
+        self.scaled_data = self.data.copy()
+        self.valid_data_array = ~np.isnan(all_data)
+        domain = self.domain
+        for attr in domain.attributes + domain.class_vars + domain.metas:
+            c = self.scaled_data.get_column_view(attr)[0]
             if attr.is_discrete:
-                no_jit[index] *= 2
-                no_jit[index] += 1
-                no_jit[index] /= 2 * len(attr.values)
+                c *= 2
+                c += 1
+                c /= 2 * len(attr.values)
```

Comment:
Isn't (2x+1)/(2a) same as (x + 0.5)/a? Aren't the above three lines the same as

```
c += 0.5
c /= len(attr.values)
```

After rewriting it this way, I also understand what the code actually does. :) Moves the data into the middle of intervals and normalizes it to [0, 1].

I know this was here before, but only God knows why it was written like this.